pkg/metapage/multi.go

```diff
@@ -0,0 +1,213 @@
+		return nil, err
+	}
+	buf := make([]byte, m.rws.PageSize()-(8*N+16))
+	if _, err := m.rws.Read(buf); err != nil {
+		return nil, err
+	}
+	// the first four bytes represents the length
+	length := binary.LittleEndian.Uint32(buf[:4])
+	return buf[4 : 4+length], nil
+}
+
+func (m *MultiPager) SetMetadata(offset uint64, data []byte) error {
+	if len(data) > m.rws.PageSize()-(8*N+16) {
+		return errors.New("metadata too large")
+	}
+	if _, err := m.rws.Seek(int64(offset)+(8*N+16), io.SeekStart); err != nil {
+		return err
+	}
+	buf := append(make([]byte, 4), data...)
+	binary.LittleEndian.PutUint32(buf, uint32(len(data)))
+	if _, err := m.rws.Write(buf); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *MultiPager) NextNOffsets(offset uint64, offsets []uint64) ([]uint64, error) {
+	if _, err := m.rws.Seek(int64(offset)+12, io.SeekStart); err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < N; i++ {
+		if err := binary.Read(m.rws, binary.LittleEndian, &offsets[i]); err != nil {
+			return nil, err
+		}
+	}
+
+	return offsets, nil
+}
+
+func (m *MultiPager) SetNextNOffsets(offset uint64, offsets []uint64) error {
+	if len(offsets) > N {
+		return fmt.Errorf("too many offsets, max number of offsets should be %d", N)
+	}
+
+	if _, err := m.rws.Seek(int64(offset)+12, io.SeekStart); err != nil {
+		return err
+	}
+
+	for _, offset := range offsets {
+		if err := binary.Write(m.rws, binary.LittleEndian, offset); err != nil {
+			return err
+		}
+	}
+
+	if err := binary.Write(m.rws, binary.LittleEndian, ^uint64(0)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *MultiPager) Next(offset uint64) (*LinkedMetaSlot, error) {
+	if _, err := m.rws.Seek(int64(offset)+12, io.SeekStart); err != nil {
+		return nil, err
+	}
+	var next btree.MemoryPointer
+	if err := binary.Read(m.rws, binary.LittleEndian, &next); err != nil {
+		return nil, err
+	}
+	return &LinkedMetaSlot{rws: m.rws, offset: next.Offset, pager: m}, nil
+}
+
+func (m *MultiPager) NextSlot(buf []byte) (int64, error) {
+	if buf != nil && len(buf) > m.rws.SlotSize() {
+		return 0, errors.New("buffer is too large")
+	}
+
+	// find next available page offset
+	for pageIndex, metaIndexUsed := range m.freeSlotIndexes {
+		for slotIndex, used := range metaIndexUsed {
+			if !used {
+				m.freeSlotIndexes[pageIndex][slotIndex] = true
+				offset := int64(pageIndex*m.rws.PageSize()) + int64(slotIndex*m.rws.SlotSize())
+				fmt.Printf("page index: %v, slot index: %v, offset: %v\n", pageIndex, slotIndex, offset)
+				return offset, nil
+			}
+		}
+	}
+
+	newPageOffset, err := m.rws.NewPage(nil)
+	if err != nil {
+		return 0, err
+	}
+
+	pageIndex := newPageOffset / int64(m.rws.PageSize())
+	metaSlotsPerPage := m.rws.PageSize() / m.rws.SlotSize()
+	metaSlotsRow := make([]bool, metaSlotsPerPage)
+	metaSlotsRow[0] = true
+	m.freeSlotIndexes = append(m.freeSlotIndexes, metaSlotsRow)
+
+	return pageIndex * int64(m.rws.PageSize()), nil
```

Comment:
I agree -- this logic should be enough for opening a new row of slots for the next subsequent page. 

I was trying to figure out why the BPtree errors out